## 이전 차시 복습 — JSON 파일 입출력

어제 내용

In [3]:
import json
import os

In [4]:
os.makedirs('sample_output', exist_ok = True)

In [5]:
messages = [
    {"role": "user",
    "content": "비밀번호를 잊어버렸어요"
    },

    {
     "role": "assistant",
     "content": "비밀번호 재설정 방법은 ~ 입니다."


    }

]

In [7]:
with open("sample_output/chat.json", 'w', encoding="utf-8") as f:
  json.dump(messages, f, ensure_ascii=False, indent = 2)

In [8]:
with open("sample_output/chat.json", 'r', encoding="utf-8") as f:
  loaded = json.load(f)

In [9]:
loaded

[{'role': 'user', 'content': '비밀번호를 잊어버렸어요'},
 {'role': 'assistant', 'content': '비밀번호 재설정 방법은 ~ 입니다.'}]

- json.dump() 함수 관련
  - messages: JSON 파일로 저장할 데이터 변수. 위 셀에서 정의된 리스트(dict 포함) 데이터가 이 변수에 담겨 있음.
  - f: 데이터를 쓸 대상 파일 객체 (위에서 as f로 지정한 파일).
  - ensure_ascii=False: 이 설정이 True(기본값)이면 한글 같은 비ASCII 문자가 \u... 형태로 변환됨. False로 설정해야 한글이 있는 그대로 파일에 저장됨.
  - indent=2: JSON 파일의 가독성을 위해 2칸 들여쓰기를 적용함. 이 설정 덕분에 파일 내용이 한 줄로 길게 늘어지지 않고 예쁘게 정렬됨.

# 2026-03-12 실습 내용 시작

In [10]:
# api key
# openai 패키지 설치
!pip install openai

In [12]:
# 코랩의 userdata 모듈을 사용하여 보안 비밀에 저장된 키를 가져옵니다.
from google.colab import userdata
import os

# 'OPENAI_API_KEY'라는 이름으로 저장한 값을 불러옵니다.
api_key = userdata.get('OPENAI_API_KEY')

# 환경 변수로 설정해두면 openai 라이브러리가 자동으로 인식하기도 합니다.
os.environ["OPENAI_API_KEY"] = api_key

print("API 키가 성공적으로 로드되었습니다.")

API 키가 성공적으로 로드되었습니다.


In [11]:
from dotenv import load_dotenv
from openai import OpenAI

load_dot_env()로 불러올 경우

#### `.env` 파일 내용

OPENAI_API_KEY = "...... 내용"

In [13]:
api_key = os.getenv("OPENAI_API_KEY")
if api_key:
  print(api_key)

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


In [14]:
client = OpenAI()

# client = OpenAI() 여기에 왜 인자로 api_key를 넣지 않아도 작동하는지?
"""
client = OpenAI()를 호출할 때 인자를 생략할 수 있었던 이유는,
OpenAI 라이브러리가 초기화될 때 시스템의 환경 변수를 자동으로 확인하도록 설계되어 있기 때문입니다.

우리는 이전 셀에서 다음과 같은 코드를 실행했습니다:
`os.environ["OPENAI_API_KEY"] = api_key`

이 코드가 OPENAI_API_KEY라는 이름의 환경 변수에 키를 등록해두었기 때문에,
OpenAI() 객체는 별도의 인자가 없어도 자동으로 그 값을 가져와 사용하게 됩니다.

"""

In [16]:
type(client)

openai.OpenAI

## 응답 구조 분석과 토큰화 (BPE)

In [30]:
# 가장 대표적인 기능인 Chat Completion 예시입니다.
  # ⭐ client.chat.completions.create(): 가장 많이 사용되는 기능으로, ChatGPT와 같은 대화형 응답을 생성합니다.
  # client.images.generate(): DALL-E를 이용해 이미지를 생성합니다.
  # client.embeddings.create(): 텍스트를 벡터(숫자 리스트)로 변환하여 검색이나 분류에 활용합니다.
  # client.audio.transcriptions.create(): 음성 파일을 텍스트로 변환(Whisper)합니다.

response = client.chat.completions.create(
    model="gpt-4o-mini", # 사용할 모델
    messages=[
        {"role": "system", "content": "당신은 친절한 어시스턴트입니다."},
        {"role": "user", "content": "OpenAI 객체가 뭔지, 어떤 기능이 있는지 항목별로 설명해주세요.(항목들만, 간단하게)"}
    ]
)

# 결과 출력
print(response.choices[0].message.content)

OpenAI 객체의 기능을 항목별로 간단하게 설명하겠습니다.

1. **자연어 처리**: 텍스트 생성, 요약, 번역 등.
2. **질문 응답**: 사용자의 질문에 대한 정보 제공.
3. **대화 생성**: 대화형 인공지능으로 사용자와의 상호작용.
4. **감정 분석**: 텍스트의 감정 판별.
5. **코드 생성 및 디버깅**: 프로그래밍 코드 작성 및 오류 수정.
6. **콘텐츠 추천**: 개인 맞춤형 콘텐츠 제안.
7. **자동화 작업**: 반복 작업의 자동 수행. 
8. **데이터 분석**: 데이터 세트에서 인사이트 도출.

이 외에도 여러 다양한 기능이 있습니다.


### 📖 OpenAI `response` 객체 구조 분석

`client.chat.completions.create()` 함수가 반환하는 객체는 **`ChatCompletion`** 타입이며, 주요 계층 구조는 다음과 같음.

#### 1. 주요 속성
- **`id`**: 요청에 대한 고유 식별자.
- **`model`**: 응답 생성에 실제 사용된 모델명.
- **`created`**: 응답이 생성된 시간(Unix 타임스탬프).
- **`choices`**: 생성된 응답 후보들의 리스트. (기본적으로 1개가 반환됨.)
- **`usage`**: 토큰 사용량 정보(입력, 출력, 합계)가 담겨 있음.

#### 2. `choices` 리스트 상세 (가장 중요)
가장 핵심적인 데이터는 `choices[0]`에 담겨 있으며, 다음과 같은 하위 속성을 가짐.
- **`.message`**: 모델이 생성한 메시지 객체.
    - **`.content`**: 실제 답변 텍스트 내용. (`print(response.choices[0].message.content)`로 출력하는 부분.)
    - **`.role`**: 메시지의 역할 (일반적으로 `assistant`).
- **`.finish_reason`**: 생성이 멈춘 이유. (예: `stop` - 정상 종료, `length` - 토큰 제한 도달)

#### 3. 코드 접근 예시
```python
# 전체 텍스트 출력
print(response.choices[0].message.content)

# 사용된 총 토큰 수 확인
print(response.usage.total_tokens)

# 종료 사유 확인
print(response.choices[0].finish_reason)
```

In [31]:
response

ChatCompletion(id='chatcmpl-DIYRmKSRTu2Y8SGN9J0xmFA28oJj5', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='OpenAI 객체의 기능을 항목별로 간단하게 설명하겠습니다.\n\n1. **자연어 처리**: 텍스트 생성, 요약, 번역 등.\n2. **질문 응답**: 사용자의 질문에 대한 정보 제공.\n3. **대화 생성**: 대화형 인공지능으로 사용자와의 상호작용.\n4. **감정 분석**: 텍스트의 감정 판별.\n5. **코드 생성 및 디버깅**: 프로그래밍 코드 작성 및 오류 수정.\n6. **콘텐츠 추천**: 개인 맞춤형 콘텐츠 제안.\n7. **자동화 작업**: 반복 작업의 자동 수행. \n8. **데이터 분석**: 데이터 세트에서 인사이트 도출.\n\n이 외에도 여러 다양한 기능이 있습니다.', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))], created=1773314690, model='gpt-4o-mini-2024-07-18', object='chat.completion', service_tier='default', system_fingerprint='fp_a1ddba3226', usage=CompletionUsage(completion_tokens=174, prompt_tokens=53, total_tokens=227, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=0, audio_tokens=0, reasoning_tokens=0, rejected_prediction_tokens=0), prompt_tokens_details=PromptTokensDeta

In [32]:
response.choices[0]

Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='OpenAI 객체의 기능을 항목별로 간단하게 설명하겠습니다.\n\n1. **자연어 처리**: 텍스트 생성, 요약, 번역 등.\n2. **질문 응답**: 사용자의 질문에 대한 정보 제공.\n3. **대화 생성**: 대화형 인공지능으로 사용자와의 상호작용.\n4. **감정 분석**: 텍스트의 감정 판별.\n5. **코드 생성 및 디버깅**: 프로그래밍 코드 작성 및 오류 수정.\n6. **콘텐츠 추천**: 개인 맞춤형 콘텐츠 제안.\n7. **자동화 작업**: 반복 작업의 자동 수행. \n8. **데이터 분석**: 데이터 세트에서 인사이트 도출.\n\n이 외에도 여러 다양한 기능이 있습니다.', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))

In [34]:
print(response.choices[0].message.content)

OpenAI 객체의 기능을 항목별로 간단하게 설명하겠습니다.

1. **자연어 처리**: 텍스트 생성, 요약, 번역 등.
2. **질문 응답**: 사용자의 질문에 대한 정보 제공.
3. **대화 생성**: 대화형 인공지능으로 사용자와의 상호작용.
4. **감정 분석**: 텍스트의 감정 판별.
5. **코드 생성 및 디버깅**: 프로그래밍 코드 작성 및 오류 수정.
6. **콘텐츠 추천**: 개인 맞춤형 콘텐츠 제안.
7. **자동화 작업**: 반복 작업의 자동 수행. 
8. **데이터 분석**: 데이터 세트에서 인사이트 도출.

이 외에도 여러 다양한 기능이 있습니다.


In [29]:
response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "system", "content": "당신은 친절한 어시스턴트입니다. 사용자의 각 질문에 대해 순서대로 번호를 매겨 한 줄씩 답변하세요."},
        {"role": "user", "content": "1. OpenAI 객체가 무엇인지, 2. response 객체의 기능이 무엇인지, 3. 가장 자주 사용되는 기능이 무엇인지 각각 설명해줘."}
    ]
)

print(response.choices[0].message.content)

1. OpenAI 객체는 OpenAI API와 상호작용하기 위한 주요 인터페이스로, 모델 호출, 설정 및 사용자 입력을 관리하는 데 사용됩니다.  
2. Response 객체는 모델이 생성한 응답을 포함하며, 텍스트, 메타데이터 및 유용한 오류 메시지를 반환하는 기능을 가지고 있습니다.  
3. 가장 자주 사용되는 기능은 생성된 텍스트를 받아오는 것이며, 이는 다양한 응용 프로그램에서 자연어 처리 작업에 활용됩니다.


In [36]:
# GPT에게 3가지 다른 질문을 해보세요.

questions = ["파이썬이 뭐야?", "llm이 뭐야?", 'api가 뭐야?']

for q in questions:
  response = client.chat.completions.create(
      model = 'gpt-4o-mini',
      messages = [

                  {"role":"user","content": q}

      ]

  )

  answer = response.choices[0].message.content
  print(answer)

파이썬(Python)은 고급 프로그래밍 언어로, 1991년 귀도 반 로썸(Guido van Rossum)에 의해 처음 개발되었습니다. 파이썬은 읽기 쉽고, 간결한 문법을 가지고 있어 배우기 쉽고, 코드 유지보수가 용이합니다. 이 언어는 다양한 프로그래밍 패러다임(객체 지향, 절차적, 함수형 등)을 지원하며, 강력한 표준 라이브러리를 제공합니다.

파이썬의 주요 특징은 다음과 같습니다:

1. **간결하고 읽기 쉬운 문법**: 코드가 명확하고 간단하여, 다른 언어보다 쉽게 이해하고 작성할 수 있습니다.
2. **다양한 라이브러리와 프레임워크**: 데이터 과학, 웹 개발, 자동화, 인공지능 등 다양한 분야에서 사용할 수 있는 많은 라이브러리와 프레임워크가 있습니다. 예를 들어, NumPy, pandas, Django, Flask 등이 있습니다.
3. **크로스 플랫폼**: 윈도우, macOS, 리눅스 등 다양한 운영체제에서 사용할 수 있습니다.
4. **활발한 커뮤니티**: 전 세계의 사용자들이 활발하게 활동하고 있어, 다양한 자료와 도움을 쉽게 찾을 수 있습니다.

이 때문에 파이썬은 웹 개발, 데이터 분석, 인공지능, 머신 러닝, 웹 스크래핑 등을 포함한 여러 분야에서 널리 사용되고 있습니다.
LLM은 "Large Language Model"의 약자로, 대규모 언어 모델을 의미합니다. 이러한 모델은 방대한 양의 텍스트 데이터를 학습하여 자연어 처리를 수행하는 인공지능 시스템입니다. LLM은 텍스트 생성, 번역, 요약, 질문 응답 등 다양한 언어 관련 작업에 사용될 수 있습니다. 대표적인 예로 OpenAI의 GPT 시리즈가 있습니다. 이러한 모델은 기계 학습 및 딥러닝 기술을 사용하여 단어와 문장의 의미를 이해하고, 인간과 유사한 방식으로 문장을 생성할 수 있습니다.
API는 "Application Programming Interface"의 약자로, 소프트웨어 애플리케이션 간의 상호작용을 가능하게 해주는 인터페이스입니다. 즉, 서로 다른 시스템이나 애플리케이션이

In [37]:
response

ChatCompletion(id='chatcmpl-DIYVlYD5xfJZ23V1URTEqRQMXhROa', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='API는 "Application Programming Interface"의 약자로, 소프트웨어 애플리케이션 간의 상호작용을 가능하게 해주는 인터페이스입니다. 즉, 서로 다른 시스템이나 애플리케이션이 서로 정보를 주고받거나 기능을 사용할 수 있도록 허용하는 규칙이나 프로토콜을 의미합니다.\n\n예를 들어, 웹사이트나 모바일 애플리케이션에서 데이터를 가져오거나 기능을 사용할 때 API를 통해 서버와 통신하게 됩니다. API는 다양한 형식으로 존재할 수 있으며, REST, SOAP, GraphQL 등이 그 예입니다.\n\nAPI를 사용하면 개발자들은 복잡한 시스템을 처음부터 만들 필요 없이 다른 시스템에서 제공하는 기능이나 데이터를 활용하여 개발 시간을 단축하고 효율성을 높일 수 있습니다.', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))], created=1773314937, model='gpt-4o-mini-2024-07-18', object='chat.completion', service_tier='default', system_fingerprint='fp_a79f70f7b4', usage=CompletionUsage(completion_tokens=165, prompt_tokens=12, total_tokens=177, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=0, audio_tokens=0, reasoning_tokens=0, rejected_predi

- role='assistant' => gpt가 대답했다는 의미
- usage=CompletionUsage(completion_tokens=165, prompt_tokens=12, total_tokens=177,   => 사용한 토큰

In [38]:
response = client.chat.completions.create(
      model = 'gpt-4o-mini',
      messages = [

                  {"role":"user","content": "대한민국의 수도는 어디인가요"}

      ]

  )

In [40]:
response.id

'chatcmpl-DIYYSS4u7WRlAJstpOR4s3GAiUFlh'

In [43]:
response.choices

[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='대한민국의 수도는 서울입니다.', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))]

In [45]:
response.choices[0].message

ChatCompletionMessage(content='대한민국의 수도는 서울입니다.', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None)

### 🏁 `finish_reason` 종류 설명

모델이 답변 생성을 중단한 이유는 `response.choices[0].finish_reason`을 통해 확인할 수 있음.

- **`stop`**: 모델이 스스로 판단하여 답변을 끝까지 완성함 (정상 종료).
- **`length`**: `max_tokens` 제한에 걸려 답변이 중간에 잘림.
- **`content_filter`**: 부적절한 콘텐츠가 감지되어 차단됨.
- **`tool_calls`**: 모델이 외부 도구(함수) 호출이 필요하다고 판단함.
- **`null`**: 생성 중 오류가 발생했거나 답변이 아직 완료되지 않음.

In [47]:
# 이번 API 호출에서 사용된 전체 토큰 수를 의미
response.usage.total_tokens

# 구체적으로 다음의 합계
  # 1. prompt_tokens: 사용자가 보낸 질문(입력)의 토큰 수
  # 2. completion_tokens: 모델이 생성한 답변(출력)의 토큰 수

23

In [ ]:
# 토큰 수는 BPE와 관련됨
# BPE : 토큰화, 텍스트를 토큰으로 바꾸는 것
# BPE (Byte Pair Encoding): GPT, Llama, Mistral 등 대다수의 생성형 AI가 선택한 방식

### 🧩 BPE(Byte Pair Encoding)란?

BPE는 텍스트를 토큰화할 때 사용하는 알고리즘으로, **단어 단위(Word-based)**와 **문자 단위(Character-based)** 토큰화의 장점을 결합한 **서브워드(Subword)** 토큰화 방식.

#### 1. 동작 방식
1.  **초기화**: 모든 문자를 개별 토큰으로 쪼갭니다. (예: `low` -> `l, o, w`)
2.  **빈도 계산**: 말뭉치에서 가장 자주 인접하게 등장하는 문자 쌍(Pair)을 찾음.
3.  **병합(Merge)**: 해당 쌍을 하나의 새로운 토큰으로 합칩니다. (예: `e`와 `s`가 자주 나오면 `es`로 합침)
4.  **반복**: 정해진 토큰 사전 크기에 도달할 때까지 이 과정을 반복함.

#### 2. BPE의 장점
-   **신조어 처리(OOV 문제 해결)**: 모르는 단어가 나와도 글자 단위로 쪼개서 처리할 수 있어 '알 수 없는 단어(Unknown Token)' 발생이 적음.
-   **효율성**: 자주 쓰이는 단어 조합은 하나의 토큰으로 묶어 데이터 압축 효과가 있음.
-   **언어 범용성**: 영어뿐만 아니라 한글, 코드 등 다양한 데이터에 적용 가능함.

#### 3. OpenAI와 BPE
OpenAI는 `tiktoken`이라는 라이브러리를 통해 개선된 BPE 알고리즘을 사용함. 이를 통해 문장을 숫자로 바꾸어 모델이 이해하게 만듭니다.

### 🧩 BPE(Byte Pair Encoding) 개념 및 수행 절차

BPE는 **단어 단위**와 **문자 단위** 토큰화의 장점을 결합한 **서브워드(Subword)** 토큰화 방식. 텍스트가 모델에 입력되기까지의 전체 과정을 정리하면 다음과 같음.

#### 1. 사전 훈련 (Training) 및 동작 원리
1.  **초기화**: 모든 문자를 개별 토큰으로 쪼갭니다. (예: `low` -> `l, o, w`)
2.  **빈도 계산**: 말뭉치에서 가장 자주 인접하게 등장하는 문자 쌍(Pair)을 찾음.
3.  **병합(Merge)**: 해당 쌍을 하나의 새로운 토큰으로 합칩니다. (예: `h`와 `e`가 자주 나오면 `he`라는 새로운 단축키 생성)
4.  **반복**: 정해진 토큰 사전 크기에 도달할 때까지 이 과정을 반복하여 수만 개의 **토큰 사전(Vocabulary)**을 완성함.

#### 2. 인코딩 및 변환 절차 (Encoding & Mapping)
1.  **매칭**: 문장이 들어오면 사전에 있는 가장 긴 단위부터 매칭하여 쪼갭니다.
    - 예: `I have` -> `I` (ID 125), `have` (ID 908)
2.  **ID 변환**: 쪼개진 각 토큰을 약속된 숫자(번호표)로 바꿉니다.
    - 결과: `[125, 908]`
3.  **임베딩 (Embedding)**: 모델은 이 숫자들을 벡터(수학적 위치)로 변환하여 계산을 시작함.

#### 3. BPE의 주요 장점
- **신조어 처리(OOV 문제 해결)**: 모르는 단어도 글자 단위로 쪼개서 처리할 수 있어 '알 수 없는 단어' 발생이 적음.
- **효율성**: 자주 쓰이는 조합은 하나의 토큰으로 묶어 데이터 압축 및 처리 속도 향상 효과가 있음.
- **언어 범용성**: 영어, 한글, 코드 등 다양한 데이터에 유연하게 적용됨.

---
**요약: `I have`라는 글자가 직접 모델에 들어가는 것이 아니라, 사전에 정의된 번호표인 `[125, 908]`이 전달되는 방식.**

In [49]:
import tiktoken

# GPT-4o 등 최신 모델이 사용하는 인코딩 방식 (o200k_base)
encoding = tiktoken.get_encoding("o200k_base")

text = "Hello, world! 안녕하세요."
tokens = encoding.encode(text)

print(f"원문: {text}")
print(f"토큰 ID 리스트: {tokens}")
print(f"토큰 개수: {len(tokens)}")

# 각 토큰 ID가 어떤 글자로 복원되는지 확인
for t in tokens:
    print(f"ID {t:6} -> {encoding.decode([t])}")

원문: Hello, world! 안녕하세요.
토큰 ID 리스트: [13225, 11, 2375, 0, 24497, 171731, 13]
토큰 개수: 7
ID  13225 -> Hello
ID     11 -> ,
ID   2375 ->  world
ID      0 -> !
ID  24497 ->  안
ID 171731 -> 녕하세요
ID     13 -> .


#### role의 종류

In [53]:
response = client.chat.completions.create(
      model = 'gpt-4o-mini',
      messages = [
          {'role':'system', 'content':'당신은 지리 선생님입니다.'},
          {"role":"user","content": "가나의 수도는 어디인가요"}

      ]

  )

response

ChatCompletion(id='chatcmpl-DIYjZorSJS3K8O8z1da23Dgwl9Cbr', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='가나의 수도는 아크라(Accra)입니다. 아크라는 가나의 가장 큰 도시이자 정치, 경제, 문화의 중심지입니다.', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))], created=1773315793, model='gpt-4o-mini-2024-07-18', object='chat.completion', service_tier='default', system_fingerprint='fp_a1ddba3226', usage=CompletionUsage(completion_tokens=35, prompt_tokens=29, total_tokens=64, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=0, audio_tokens=0, reasoning_tokens=0, rejected_prediction_tokens=0), prompt_tokens_details=PromptTokensDetails(audio_tokens=0, cached_tokens=0)))

### 🎭 `role`의 종류 및 역할

- **`system`**: 모델의 페르소나나 행동 지침을 설정함. (예: "당신은 친절한 수학 선생님.")
- **`user`**: 사용자가 모델에게 보내는 질문이나 명령.
- **`assistant`**: 모델(AI)이 이전에 했던 대답. 대화의 맥락을 유지하기 위해 사용자가 직접 과거 대화 내역을 넣을 때 사용함.
- **`tool`** (또는 `function`): 모델이 특정 외부 도구(함수)를 호출한 결과값을 전달할 때 사용함.

In [62]:
response = client.chat.completions.create(
      model = 'gpt-4o-mini',
      messages = [
          {'role':'system', 'content':'당신은 10년차 상담원입니다. 상냥한 답변을 제공해야합니다. 이모지를 최대한 이용해주세요.'},
          {"role":"user","content": "비밀번호를 잊어버렸어요."}

      ]

  )
response.choices[0].message.content

'아이고, 비밀번호를 잊어버리셨군요! 😅 걱정 마세요, 도움을 드릴게요! 💪\n\n대부분의 서비스에서는 비밀번호 재설정 방법이 있습니다. 웹사이트나 앱의 로그인 화면에 보시면 “비밀번호 찾기” 또는 “비밀번호 재설정” 옵션이 있을 거예요. ✨ 그걸 클릭하신 후, 지침을 따라 주시면 좋습니다! \n\n이 과정에서 어려움이 있으시면 언제든지 말씀해 주세요! 📩💕'

In [61]:
# 서로 다른 2개의 시스템 메시지로 gpt 성격을 바꿔서 답변을 출력해주세요.

In [64]:
response = client.chat.completions.create(
      model = 'gpt-4o-mini',
      messages = [
          {'role':'system', 'content':'당신은 10년차 프로그래머입니다. 최대한 전문적이고 효과적인 정보를 간결하게 제공해야합니다.'},
          {"role":"user","content": "비밀번호를 잊어버렸어요."}

      ]

  )
print(response.choices[0].message.content)

비밀번호를 잊어버렸다면, 다음 단계를 따라 복구해 보세요:

1. **비밀번호 찾기 링크 사용**: 로그인 화면에서 "비밀번호를 잊으셨나요?" 또는 유사한 링크를 클릭하세요.
2. **이메일 주소 입력**: 계정에 등록된 이메일 주소를 입력합니다.
3. **이메일 확인**: 비밀번호 재설정 링크가 포함된 이메일을 확인하고, 링크를 클릭합니다.
4. **새 비밀번호 설정**: 안내에 따라 새 비밀번호를 설정합니다.

이메일을 찾을 수 없는 경우, 고객 지원에 문의하거나 추가적인 인증 절차를 따라야 할 수 있습니다.


## 모델 파라미터 제어

#### temperature

In [67]:
question = '인공지능의 미래에 대해 한 문장으로 말해주세요.'
response = client.chat.completions.create(
    model = 'gpt-4o-mini', # gpt-5-mini를 gpt-4o-mini로 수정
    messages = [{'role':'user', 'content':question}],
    temperature = 0
)

print(response.choices[0].message.content)

인공지능의 미래는 인간의 삶을 혁신적으로 변화시키며, 다양한 분야에서 협력과 창의성을 증진시키는 방향으로 발전할 것입니다.


In [78]:
question = '인공지능의 미래에 대해 한 문장으로 말해주세요.'
response = client.chat.completions.create(
    model = 'gpt-4o-mini', # gpt-5-mini를 gpt-4o-mini로 수정
    messages = [{'role':'user', 'content':question}],
    temperature  = 1.95
)

print(response.choices[0].message.content)

인공지능의 발전은 소 dəyişikrews خلاص خود بهذه预 تطوير perguntas, educativa	respделанию músicasisierte оборудование】【。


In [79]:
import numpy as np

def calculate_softmax(logits, T):
  z = np.array(logits)
  e_z = np.exp(z / T-np.max(z/T))
  return e_z / e_z.sum()

In [80]:
logits = [2.0, 1.5, 1.0, 0.5]
temperature = [0.1, 0.5, 1.0, 2.0, 10.0, 100]

In [86]:
for T in temperature:
  probs = calculate_softmax(logits, T)
  print(probs)

[9.93262055e-01 6.69254708e-03 4.50940275e-05 3.03841168e-07]
[0.64391426 0.23688282 0.08714432 0.0320586 ]
[0.45505423 0.27600434 0.1674051  0.10153632]
[0.34993201 0.27252732 0.21224449 0.16529618]
[0.26905047 0.25592872 0.24344693 0.23157388]
[0.25187811 0.25062187 0.24937188 0.24812814]


### 🌡️ Temperature(온도) 설정의 이해

Temperature는 모델의 **창의성**과 **무작위성**을 조절하는 매개변수. 내부적으로는 **Softmax** 함수에 들어가기 전 값(Logits)을 나누는 역할을 함.

#### 1. 온도가 낮을 때 (`T -> 0`)
- **확률 분포**: 가장 높은 점수를 가진 토큰의 확률이 압도적으로 높아짐.
- **결과**: 매우 결정론적(Deterministic)이고 보수적인 답변이 나옴. 사실 기반의 질문(예: 수도가 어디야?)에 적합함.
- **수식 효과**: `logits / 0.1`과 같이 작은 수로 나누면 값들 사이의 격차가 매우 커짐.

#### 2. 온도가 기본일 때 (`T = 1.0`)
- **확률 분포**: 모델이 학습한 원래의 확률 분포를 그대로 사용함.
- **결과**: 적절한 창의성과 논리적 흐름을 동시에 유지함.

#### 3. 온도가 높을 때 (`T > 1.0`)
- **확률 분포**: 토큰들 간의 확률 차이가 완만해짐(Flattening).
- **결과**: 평소에 잘 선택하지 않던 의외의 단어를 선택하여 답변이 창의적이고 다양해짐. 하지만 너무 높으면 논리가 깨지거나 무의미한 문장이 생성됨.

---
**💡 요약: 정확한 정보가 필요할 땐 `0`, 창의적인 아이디어가 필요할 땐 `0.7~1.0` 사이를 추천함.**

### 🎲 Top-p 및 Top-k 샘플링

Temperature 외에도 토큰 선택의 다양성을 조절하는 중요한 기법들.

#### 1. Top-p (Nucleus Sampling)
- **동작**: 누적 확률이 설정값(p)에 도달할 때까지 상위 후보 토큰들을 포함함.
- **특징**: 확률 분포의 '핵심(Nucleus)'만 골라내는 방식으로, 상황에 따라 선택되는 후보군의 개수가 유동적으로 변함.
- **예**: `p=0.9`이면 상위 90%의 확률을 차지하는 토큰들 중에서만 다음 단어를 고름.

#### 2. Top-k
- **동작**: 확률이 가장 높은 상위 k개의 토큰 후보 중에서만 다음 단어를 선택함.
- **특징**: 후보군의 개수를 고정적으로 제한하여 전혀 엉뚱한 단어가 나오는 것을 방지함.
- **예**: `k=50`이면 확률 순위 1~50위 안의 토큰만 고려함.

---
**💡 팁: OpenAI API에서는 주로 `Temperature`와 `Top-p` 중 하나만 조절하는 것을 권장하며, 둘 다 조절할 경우 결과 예측이 어려워질 수 있음.**

In [ ]:
# reasoning 모델의 경우는 top-p, top-k는 고정함
"""
OpenAI의 o1(Reasoning) 모델과 같은 추론 전용 모델들은 일반적인 Chat 모델과는 설정 방식이 조금 다릅니다.

샘플링 제한: o1 모델들은 내부적인 '생각(Reasoning)' 과정을 거쳐 최적의 논리적 결론을 내야 합니다. 이때 temperature, top_p, presence_penalty 같은 값을 조절하면 추론 과정이 불안정해질 수 있어, 대부분 1.0(기본값)으로 고정되거나 수정을 권장하지 않습니다.
모델의 특징: 창의성보다는 정확한 논리가 중요하기 때문에 무작위성을 부여하는 파라미터 사용이 제한적인 것입니다.

"""

In [88]:
question = "고양이를 한 문장으로 소개해줘,"

for temp in [0, 0.5, 1.0]:
  print(f"--- 현재 온도: {temp} ---")
  for i in range(2):
    response = client.chat.completions.create(
        model = 'gpt-4o-mini',
        messages = [{'role':'user', 'content':question}],
        temperature = temp
    )
    # print문을 들여쓰기하여 루프 안으로 이동
    print(f"{i} : {response.choices[0].message.content}")

--- 현재 온도: 0 ---
0 : 고양이는 독립적이고 호기심 많은 성격을 가진 귀여운 반려동물로, 부드러운 털과 매력적인 눈빛으로 많은 사람들에게 사랑받고 있습니다.
1 : 고양이는 독립적이고 호기심 많은 성격을 가진 귀여운 반려동물로, 부드러운 털과 매력적인 눈빛으로 많은 사람들에게 사랑받고 있습니다.
--- 현재 온도: 0.5 ---
0 : 고양이는 독립적이고 호기심 많은 성격을 가진 귀여운 반려동물로, 부드러운 털과 매력적인 눈빛으로 많은 사랑을 받습니다.
1 : 고양이는 독립적이고 호기심 많은 성격을 가진 귀여운 반려동물로, 부드러운 털과 매력적인 눈빛으로 사람들의 사랑을 받습니다.
--- 현재 온도: 1.0 ---
0 : 고양이는 독립적이고 호기심 많은 성격을 가진, 부드러운 털과 우아한 움직임이 특징인 사랑스러운 반려동물입니다.
1 : 고양이는 독립적이고 호기심 많은 성격을 가진 귀여운 반려동물로, 부드러운 털과 민첩한 몸놀림이 특징입니다.


#### max_token

In [94]:
#### max_token
response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':'대한민국에 대해 알려주세요.'}],
    max_tokens=30
)

In [96]:
response
# finish_reason ='length' 임
# 중간에 답변이 짤림


ChatCompletion(id='chatcmpl-DIZQ1pzlXYga5jIQ9ps0zMbugZ5Hq', choices=[Choice(finish_reason='length', index=0, logprobs=None, message=ChatCompletionMessage(content='대한민국, 흔히 한국이라고도 불리는 이 나라는 동아시아에 위치한 국가로, 한반도의 남쪽에 자리 잡', refusal=None, role='assistant', annotations=[], audio=None, function_call=None, tool_calls=None))], created=1773318425, model='gpt-4o-mini-2024-07-18', object='chat.completion', service_tier='default', system_fingerprint='fp_a1ddba3226', usage=CompletionUsage(completion_tokens=30, prompt_tokens=14, total_tokens=44, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=0, audio_tokens=0, reasoning_tokens=0, rejected_prediction_tokens=0), prompt_tokens_details=PromptTokensDetails(audio_tokens=0, cached_tokens=0)))

In [103]:
question = "대한민국에 대해 알려주세요."

for temp in [0, 0.5, 1.0]:
  print(f"--- 현재 온도: {temp} ---")
  for token in [10, 50, 100]:
    print(f"  ---- 현재 토큰 수: {token} ---")
    response = client.chat.completions.create(
        model = 'gpt-4o-mini',
        messages = [{'role':'user', 'content':question}],
        temperature = temp,
        max_tokens = token
    )
    print(f"{i} : {response.choices[0].message.content}")

--- 현재 온도: 0 ---
  ---- 현재 토큰 수: 10 ---
1 : 대한민국(大韓民國), 흔히
  ---- 현재 토큰 수: 50 ---
1 : 대한민국(大韓民國), 흔히 한국이라고 불리는 나라는 동아시아에 위치한 국가로, 한반도의 남쪽에 자리잡고 있습니다. 북쪽으로는 북한(조선민주주의인민공
  ---- 현재 토큰 수: 100 ---
1 : 대한민국(大韓民國), 흔히 한국이라고 불리는 나라는 동아시아에 위치한 국가로, 한반도의 남쪽에 자리잡고 있습니다. 북쪽으로는 북한과 접하고 있으며, 서쪽은 황해, 동쪽은 동해, 남쪽은 대한해협과 접해 있습니다. 

### 기본 정보
- **수도**: 서울
- **면적**: 약 100,210 평방킬로미터

--- 현재 온도: 0.5 ---
  ---- 현재 토큰 수: 10 ---
1 : 대한민국(大韓民國), 흔히
  ---- 현재 토큰 수: 50 ---
1 : 대한민국(大韓民國), 흔히 한국이라고 불리는 나라는 동아시아에 위치한 국가로, 한반도의 남쪽에 있습니다. 북쪽으로는 북한과 국경을 접하고 있으며, 서쪽으로는
  ---- 현재 토큰 수: 100 ---
1 : 대한민국, 또는 한국은 동아시아에 위치한 국가로, 한반도의 남쪽 부분을 차지하고 있습니다. 북쪽에는 북한이 있으며, 동쪽으로는 일본과의 해협, 서쪽으로는 중국과의 해협이 있습니다. 대한민국의 수도는 서울이며, 주요 도시로는 부산, 인천, 대구, 대전 등이 있습니다.

### 역사
대한민국은 1945년 제2차 세계대전의 종전 이후
--- 현재 온도: 1.0 ---
  ---- 현재 토큰 수: 10 ---
1 : 대한민국(大韓民國), 줄여
  ---- 현재 토큰 수: 50 ---
1 : 대한민국, 또는 한국은 동아시아에 위치한 국가로, 한반도의 남부에 위치하고 있습니다. 북쪽으로는 조선민주주의인민공화국(북한)과 국경을 접하고 있으며
  ---- 현재 토큰 수: 100 ---
1 : 대한민국, 흔히 한국이라고 불리는 이 나라는 동아시아의 한반도에 위치하며, 북쪽

### 🧩 Penalty 설정 (frequency_penalty, presence_penalty)

답변 내에서 단어의 반복을 억제하고 다양성을 높이는 설정.

#### 1. frequency_penalty (빈도 페널티)
- 같은 단어가 반복해서 나오지 못하게 조절함.
- 0: 반복 허용, 1~2: 반복 억제

#### 2. presence_penalty (존재 페널티)
- 이미 등장한 단어보다는 새로운 단어나 주제를 선택하도록 유도함.
- 0: 기존 주제 유지, 1~2: 적극적으로 새로운 내용 생성

---
**💡 요약: 반복을 피하고 싶다면 `0.1~0.5` 사이의 값을 추천함.**

In [105]:
# top_p
question = "봄에 대한 짧은 시를 써주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    top_p = 0.1
)

print(f"{response.choices[0].message.content}")

봄바람 살랑, 꽃망울 터지고  
햇살 아래 새싹이 고개를 내밀어  
푸른 잎사귀, 노래하는 새들  
희망의 향기, 세상에 퍼지네  

따스한 날씨, 마음도 풀어져  
기다림 끝에 찾아온 기쁨  
봄은 다시, 생명의 시작  
우리의 꿈도 함께 피어나네.


In [107]:
question = "봄에 대한 짧은 시를 써주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    top_p = 0.95
)

print(f"{response.choices[0].message.content}")

봄의 속삭임

새싹이 움트고, 꽃이 피어나  
햇살에 반짝이는 나른한 오후,  
바람은 부드럽고, 마음은 가벼워  
사랑의 노래가 들려오는 계절.

하늘은 푸르고, 구름은 유유히  
길가에 핀 벚꽃, 그리움의 향기,  
우리의 손을 맞잡고 나가보자,  
봄의 꿈속으로, 함께 걸어가자.


In [111]:
# penalty

question = "커피에 대해 다섯 문장으로 설명해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    frequency_penalty=0.0
)

print(f"{response.choices[0].message.content}")

커피는 커피나무의 열매인 커피 체리를 가공하여 만든 음료입니다. 전 세계적으로 많은 사람들이 즐겨 마시는 음료로, 카페인 성분이 있어 각성 효과가 있습니다. 원두 종류에 따라 맛과 향이 다르며, 대표적으로 아라비카와 로부스타가 있습니다. 커피는 에스프레소, 아메리카노, 라떼 등 다양한 형태로 즐길 수 있습니다. 또한, 커피 문화는 각 나라마다 다르게 발전하여 독특한 음료 및 음용 방법이 존재합니다.


In [116]:
# frequency_penalty

question = "커피에 대해 다섯 문장으로 설명해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    frequency_penalty=2 # 높을수록 이미 썼던 단어를 반복하지 않으려고 함
)

print(f"{response.choices[0].message.content}")

커피는 커피나무의 열매에서 얻은 원두를 로스팅하여 만든 음료입니다. 전 세계적으로 인기가 높으며, 아침 식사 또는 휴식 시간에 자주 즐겨 마십니다. 다양한 종류와 블렌드가 있으며, 에스프레소, 아메리카노, 카푸치노 등 여러 형태로 제공됩니다. 카페인 성분이 포함되어 있어 피로 회복과 집중력 향상에 도움을 줄 수 있습니다. 또한 커피는 사회적 상호작용의 장으로 자리잡아 많은 사람들이 커피숍에서 모임을 가지기도 합니다.


In [118]:
# presence_penalty

question = "커피에 대해 다섯 문장으로 설명해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    presence_penalty=0.0
)

print(f"{response.choices[0].message.content}")

커피는 커피나무의 씨앗인 커피콩을 원료로 하여 만들어진 음료입니다. 주로 아프리카, 아시아, 남미에서 재배되며, 다양한 종류와 맛을 지니고 있습니다. 커피는 카페인을 포함하고 있어, 섭취 시 각성 효과를 주어 피로 회복에 도움을 줍니다. 일반적으로 에스프레소, 아메리카노, 카페 라떼 등 여러 가지 방식으로 조리됩니다. 많은 사람들에게는 일상적인 음료이자 사회적 상호작용의 중요한 요소로 자리 잡고 있습니다.


In [119]:
# presence_penalty

question = "커피에 대해 다섯 문장으로 설명해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    presence_penalty=2 # 높을수록 새로운 주제를 생성할 확률이 높아짐.
)

print(f"{response.choices[0].message.content}")

커피는 커피나무의 열매에서 추출한 음료로, 전 세계적으로 인기 있는 음료 중 하나입니다. 주로 카페인을 포함하고 있어 각성 효과가 있으며, 피로 회복과 집중력 향상에 도움을 줍니다. 커피는 다양한 방법으로 제조되며, 에스프레소, 드립 커피, 프렌치 프레스 등 여러 종류가 있습니다. 지역마다 특색 있는 맛과 향이 다르며, 원두의 품질과 로스팅 방법에 따라 풍미가 달라집니다. 최근에는 커피 문화가 발전하면서 스페셜티 커피와 커피 관련 페어링 등이 대중의 관심을 받고 있습니다.


### ⏹️ stop 및 🪙 seed 설정

#### 1. stop (중단 시퀀스)
- 특정 단어가 등장하면 즉시 답변 생성을 멈춤.
- 예: `stop="."` (첫 마침표에서 종료)

#### 2. seed (시드값)
- 답변의 무작위성을 제어하여 동일한 질문에 대해 최대한 일관된 답변을 출력하게 함.
- 예: `seed=42` (동일한 시드 사용 시 결과 재현성 향상)

---
**💡 팁: 특정 형식(예: 번호 매기기)을 강제하거나, 실험의 일관성을 유지할 때 유용함.**

In [120]:
# stop, seed

question = "프로그래밍 언어 10개를 번호로 나열해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    stop = ["6."]
)

print(f"{response.choices[0].message.content}")

물론입니다! 다음은 프로그래밍 언어 10개의 목록입니다:

1. Python
2. Java
3. JavaScript
4. C++
5. C#



In [121]:
# stop, seed

question = "프로그래밍 언어 10개를 번호로 나열해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
)

print(f"{response.choices[0].message.content}")

물론입니다! 아래는 10개의 프로그래밍 언어를 번호로 나열한 것입니다.

1. 파이썬 (Python)
2. 자바 (Java)
3. 자바스크립트 (JavaScript)
4. C++
5. C#
6. 루비 (Ruby)
7. PHP
8. 스위프트 (Swift)
9. 고 (Go)
10. 타입스크립트 (TypeScript) 

이 외에도 많은 프로그래밍 언어들이 있습니다!


In [122]:
# seed : 시드값이 동일하면 비슷한 답변을 생성

question = "사과에 대해 설명해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    seed = 42
)

print(f"{response.choices[0].message.content}")

사과는 과일의 한 종류로, 주로 사과나무(학명: Malus domestica)에서 자생하는 열매입니다. 사과는 전 세계적으로 소비되는 가장 인기 있는 과일 중 하나로, 다양한 품종이 존재합니다. 

### 사과의 특징
- **모양과 색상**: 일반적으로 원형 또는 타원형을 가지며, 색상은 빨강, 초록, 노랑 등 다양합니다.
- **향과 맛**: 상큼하고 달콤한 맛이 특징이며, 품종에 따라 신맛도 느낄 수 있습니다.
- **영양 가치**: 사과는 비타민 C, 식이섬유, 항산화 물질이 풍부하여 건강에 이로운 과일로 알려져 있습니다. 특히, 사과의 껍질에 많은 영양소가 포함되어 있습니다.

### 건강 효능
사과에는 여러 가지 건강 효능이 있습니다. 주요 효능으로는:
- **소화 개선**: 식이섬유가 풍부하여 장 건강에 도움을 줍니다.
- **면역력 강화**: 비타민 C가 다량 포함되어 있어 면역 체계를 강화하는 데 도움이 됩니다.
- **심혈관 건강**: 항산화 물질이 풍부하여 심혈관 건강에 긍정적인 영향을 미친다고 알려져 있습니다.
- **체중 관리**: 칼로리가 낮고 포만감을 주어 다이어트에 도움이 될 수 있습니다.

### 요리 활용
사과는 생으로 먹거나, 요리에 다양한 방법으로 활용될 수 있습니다. 사과 파이, 잼, 주스, 샐러드 등에 사용되며, 샐러드나 고기 요리에 곁들여지기도 합니다.

### 문화적 의미
사과는 여러 문화에서 상징적인 의미를 갖고 있습니다. 예를 들어, 유혹이나 지혜의 상징으로 여겨지기도 하며, 다양한 신화와 전통에서 중요한 역할을 합니다.

사과는 그 자체로 맛있고 건강한 간식이 될 뿐만 아니라, 다채로운 요리로 활용될 수 있는 과일입니다.


In [124]:
# seed

question = "사과에 대해 설명해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    seed = 42
)

print(f"{response.choices[0].message.content}")

사과는 과일의 한 종류로, 주로 사과나무(학명: Malus domestica)에서 자생하는 열매입니다. 사과는 전 세계적으로 소비되는 가장 인기 있는 과일 중 하나로, 다양한 품종이 존재합니다. 

### 사과의 특징
- **모양과 색상**: 일반적으로 원형 또는 타원형을 가지며, 색상은 빨강, 초록, 노랑 등 다양합니다.
- **향과 맛**: 상큼하고 달콤한 맛이 특징이며, 품종에 따라 신맛도 느낄 수 있습니다.
- **영양 가치**: 사과는 비타민 C, 식이섬유, 항산화 물질이 풍부하여 건강에 이로운 과일로 알려져 있습니다. 특히, 피부 건강과 소화에 도움을 줄 수 있습니다.

### 사과의 용도
- **생으로 섭취**: 과일로서 직접 먹거나 샐러드에 첨가됩니다.
- **요리**: 사과 파이, 사과 소스, 잼 등 다양한 요리에 사용됩니다.
- **음료**: 사과 주스나 사과 식초로 가공되어 소비됩니다.

### 문화적 의미
사과는 여러 문화에서 상징적인 의미를 가지고 있습니다. 예를 들어, 고대 그리스 신화에서 사과는 아름다움과 사랑을 상징하며, ‘금사과’ 이야기가 유명합니다. 현대에서는 사과가 지혜와 학습의 상징으로 여겨지는 경우도 많습니다.

사과는 건강에 좋은 과일일 뿐만 아니라, 다양한 요리와 문화적 의미를 지니고 있어 많은 사람들에게 사랑받고 있습니다.


In [125]:
# seed

question = "사과에 대해 설명해주세요."


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    seed = 11
)

print(f"{response.choices[0].message.content}")

사과는 과일 중 하나로, 농작물로 재배되는 사과나무(학명: Malus domestica)의 열매입니다. 사과는 전 세계적으로 널리 소비되며, 다양한 품종과 맛이 있습니다. 일반적으로 사과는 달콤하거나 시큼하며, 아삭한 식감이 특징입니다.

### 사과의 특징
1. **영양가**: 사과는 비타민 C, 식이섬유, 항산화 물질 등이 풍부하여 건강에 이로운 과일입니다. 특히, 식이섬유는 소화 건강에 도움을 줍니다.
2. **다양한 품종**: 사과는 여러 가지 품종이 있으며, 각 품종마다 색깔, 크기, 맛이 다릅니다. 인기 있는 품종으로는 빨간 사과의 '홍옥', 노란 사과의 '골든 딜리셔스', 푸른 사과의 '그라니 스미스' 등이 있습니다.
3. **사용 용도**: 사과는 생식으로 먹는 것뿐만 아니라, 주스, 잼, 파이, 샐러드 등 다양한 요리에도 활용됩니다. 또한, 사과 식초로도 사용되며, 화장품과 건강 보조 식품에도 포함됩니다.
4. **재배**: 사과나무는 온대 기후에서 잘 자라며, 일반적으로 가을에 수확합니다. 적절한 관리와 기후가 맞아야 좋은 품질의 사과를 수확할 수 있습니다.

### 건강 효과
사과는 여러 건강 효과로 알려져 있습니다. 심혈관 건강, 체중 관리, 혈당 조절에 도움을 줄 수 있으며, 특정 질병 예방에도 긍정적인 영향을 미칠 수 있습니다.

사과는 그 맛과 다양성 때문에 많은 사람들이 즐겨 찾는 과일이며, 전 세계적으로 농업의 중요한 작물 중 하나입니다.


In [136]:
# stop, seed 이용해 과일 10개를 나열해주세요
question = "과일 종류 10개를 나열해주세요"


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    stop = "5",
    seed = 636
)

print(f"{response.choices[0].message.content}")

다음은 다양한 과일 종류 10개입니다:

1. 사과
2. 바나나
3. 오렌지
4. 포도



In [138]:
# stop, seed 이용해 과일 10개를 나열해주세요
question = "과일 종류 10개를 나열해주세요"


response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
    stop = "15",
    seed = 42
)

print(f"{response.choices[0].message.content}")

물론입니다! 다음은 다양한 과일 종류 10개입니다:

1. 사과
2. 바나나
3. 오렌지
4. 포도
5. 딸기
6. 수박
7. 파인애플
8. 키위
9. 블루베리
10. 자몽

이 외에도 많은 종류의 과일이 있습니다!


## 멀티턴 대화와 컨텍스트 관리

### 🧠 컨텍스트(Context) 기억 기능

OpenAI 모델은 기본적으로 **'상태가 없는(Stateless)'** 방식. 즉, 이전 대화를 스스로 기억하지 못하므로 개발자가 직접 대화 내역을 전달해야 함.

#### 1. 작동 원리
- 새로운 질문을 보낼 때, **이전 질문(user)**과 **이전 답변(assistant)**을 모두 `messages` 리스트에 포함하여 다시 전송함.
- 모델은 전달받은 전체 대화 내역을 읽고 문맥을 파악하여 답변함.

#### 2. 메시지 구성 방식
- `system`: 대화의 전체적인 지침 설정
- `user`: 사용자의 현재 및 과거 질문들
- `assistant`: 모델이 이전에 내놓았던 답변들 (맥락 유지를 위해 반드시 포함)

---
**💡 주의: 대화가 길어질수록 전송하는 토큰 수가 늘어나 비용이 증가하므로, 필요에 따라 과거 대화의 일부만 유지하거나 요약하는 전략이 필요함.**

In [168]:
question = "비밀번호를 잊어버렸어요"

messages = []

response = client.chat.completions.create(
    model = 'gpt-4o-mini',
    messages = [{'role':'user', 'content':question}],
)

answer1 = response.choices[0].message.content
print(answer1)

비밀번호를 잊어버린 경우 다음 단계를 따라 보세요:

1. **비밀번호 재설정 옵션 찾기**: 로그인 화면에서 "비밀번호를 잊으셨나요?" 또는 "비밀번호 재설정" 링크를 찾아 클릭하세요.

2. **이메일 또는 전화번호 입력**: 해당 계정에 등록된 이메일 주소나 전화번호를 입력합니다.

3. **재설정 링크 확인**: 입력한 이메일이나 문자 메시지로 비밀번호 재설정 링크가 전송됩니다. 해당 링크를 클릭하세요.

4. **새 비밀번호 설정**: 링크를 통해 새 비밀번호를 설정하고 확인하세요.

5. **새 비밀번호로 로그인**: 새 비밀번호로 로그인해 보세요.

만약 이 방법으로도 문제가 해결되지 않는다면, 고객 지원 센터에 문의하여 추가 도움을 요청할 수 있습니다. 


In [169]:
# 3턴 이상의 대화 생성

messages = [
    {'role':'system', 'content':'당신은 친절한 AI입니다. 짧게 답변합니다.'}
]

question = ["내 이름은 홍길동이야. "]



In [170]:
# 1턴: 이름 알려주기
question1 = "내 이름은 홍길동이야."
messages.append({'role': 'user', 'content': question1})

response = client.chat.completions.create(
    model='gpt-4o-mini',
    messages=messages
)

answer1 = response.choices[0].message.content
messages.append({'role': 'assistant', 'content': answer1})
print(f"사용자: {question1}")
print(f"AI: {answer1}\n")

사용자: 내 이름은 홍길동이야.
AI: 안녕하세요, 홍길동님! 만나서 반갑습니다. 어떻게 도와드릴까요?



In [171]:
# 2턴: 이름을 기억하는지 확인하기
question2 = "내 이름이 뭐라고?"
messages.append({'role': 'user', 'content': question2})

response = client.chat.completions.create(
    model='gpt-4o-mini',
    messages=messages
)

answer2 = response.choices[0].message.content
messages.append({'role': 'assistant', 'content': answer2})
print(f"사용자: {question2}")
print(f"AI: {answer2}\n")

사용자: 내 이름이 뭐라고?
AI: 홍길동님이라고 하셨습니다!



In [172]:
# 3턴: 추가 대화 나누기
question3 = "오늘 날씨 어때?"
messages.append({'role': 'user', 'content': question3})

response = client.chat.completions.create(
    model='gpt-4o-mini',
    messages=messages
)

answer3 = response.choices[0].message.content
messages.append({'role': 'assistant', 'content': answer3})
print(f"사용자: {question3}")
print(f"AI: {answer3}")

사용자: 오늘 날씨 어때?
AI: 현재 날씨 정보는 제공할 수 없지만, 기상 앱이나 웹사이트에서 확인해 보시면 좋습니다!


In [176]:
response = client.chat.completions.create(
    model = 'gpt-4o-mini', messages = [{'role':'user','content':'파이썬의 장점 3가지를 설명해줘'}],
    stream = True

)

full_answer = ""

for chunk in response:
  text = chunk.choices[0].delta.content
  if text:
    print(text, end="",flush=True)
    full_answer += text


파이썬의 장점은 다양하지만, 그 중에서 특히 중요한 세 가지를 소개하겠습니다.

1. **가독성**: 파이썬은 문법이 간단하고 명확하여 코드를 쉽게 읽고 이해할 수 있습니다. 이는 개발자들이 코드 작성과 유지보수를 더 효율적으로 할 수 있도록 도와줍니다. 특히, 들여쓰기를 기반으로 한 구조는 코드의 블록을 명확하게 구분하므로 팀 내 협업이 용이합니다.

2. **풍부한 라이브러리와 생태계**: 파이썬은 데이터 과학, 인공지능, 웹 개발, 자동화 등 다양한 분야에서 활용될 수 있는 풍부한 라이브러리와 프레임워크를 갖추고 있습니다. 예를 들어, NumPy와 Pandas는 데이터 분석에, TensorFlow와 PyTorch는 머신러닝에, Django와 Flask는 웹 개발에 유용합니다. 이러한 생태계 덕분에 개발자들은 필요한 도구를 쉽게 찾아 사용할 수 있습니다.

3. **플랫폼 독립성**: 파이썬은 다양한 운영체제에서 실행될 수 있습니다. 윈도우, 리눅스, macOS 등에서 동일한 코드가 거의 변경 없이 작동하기 때문에,_cross-platform 개발이 용이합니다. 이는 개발자들이 특정 플랫폼에 종속되지 않고 다양한 환경에서 애플리케이션을 배포할 수 있게 해줍니다.

이 외에도 파이썬은 대화형 프로그래밍 환경, 대규모 커뮤니티 지원 등 많은 장점을 가지고 있어, 초보자뿐만 아니라 경험이 많은 개발자들에게도 널리 사용되고 있습니다.